include/acacia/atspi/atspi_interface.h

```diff
@@ -0,0 +1,51 @@
+#ifndef INCLUDE_ACACIA_ATSPI_ATSPI_INTERFACE_H_
+#define INCLUDE_ACACIA_ATSPI_ATSPI_INTERFACE_H_
+
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include <atspi/atspi.h>
+
+namespace acacia {
+
+/**
+ * Objects descending from this class wrap an
+ * [AtspiObject](https://docs.gtk.org/atspi2/class.Object.html) of a specific
+ * type. See descendants of this class for more information about each type.
+ */
+template <typename T>
+class AtspiInterface {
+ public:
+  explicit AtspiInterface(T* interface) : interface_(interface, Deleter()) {}
+
+  AtspiInterface() {}
+  virtual ~AtspiInterface() {}
+
+  /**
+   * Tests whether the underlying interface pointer is null.
+   */
+  bool isNull() const { return !interface_; }
+
+  /**
+   * Helper function to print commonly needed values associated with this
+   * interface.
+   */
+  virtual std::string toString() const = 0;
+
+ protected:
+  class Deleter {
+   public:
+    void operator()(T* ptr) {
+      if (ptr == nullptr)
+        return;
+      g_object_unref(ptr);
+    }
+  };
```

Comment:
The `Deleter` class? I learned it from the same [heroic PR](https://github.com/Igalia/acacia/pull/171/files#diff-6acaa5ccb23aa8579f65ac9c02d9e6668713c18a2db32ed7ab9f24d625043606R13) - I just pulled the lambda expression (!!) into an explicit nested class, because my brain can only cope with so much syntax.

Reply:
yeah the deleter argument to the shared_ptr. Nice :)

But speaking of syntax, it definitely took me a moment to understand what this pattern of `class Deleter { public: void operator() ...` was about.